### PR TITLE
Fix boundary norm negative

### LINKED ISSUE #21671: [Bug]: 3.5.0 colorbar ValueError: minvalue must be less than or equal to maxvalue
### Bug summary

ValueError is raised on colorbar code that worked with previous matplotlib releases.

### Code for reproduction

```python
import matplotlib.pyplot as plt
import matplotlib.cm as mpcm
import matplotlib.colors as mpcolors

cmap = plt.get_cmap("viridis")
clevs = list(range(-94, -85))
norm = mpcolors.BoundaryNorm(clevs, cmap.N)
plt.colorbar(
    mpcm.ScalarMappable(cmap=cmap, norm=norm),
)
```

### Actual outcome

```
Traceback (most recent call last):
  File "colorbar.py", line 8, in <module>
    plt.colorbar(
  File "/opt/miniconda3/envs/prod/lib/python3.8/site-packages/matplotlib/pyplot.py", line 2088, in colorbar
    ret = gcf().colorbar(mappable, cax=cax, ax=ax, **kw)
  File "/opt/miniconda3/envs/prod/lib/python3.8/site-packages/matplotlib/figure.py", line 1158, in colorbar
    cb = cbar.Colorbar(cax, mappable, **cb_kw)
  File "/opt/miniconda3/envs/prod/lib/python3.8/site-packages/matplotlib/colorbar.py", line 493, in __init__
    self.draw_all()
  File "/opt/miniconda3/envs/prod/lib/python3.8/site-packages/matplotlib/colorbar.py", line 577, in draw_all
    X, Y, extendlen = self._mesh()
  File "/opt/miniconda3/envs/prod/lib/python3.8/site-packages/matplotlib/colorbar.py", line 1132, in _mesh
    y, extendlen = self._proportional_y()
  File "/opt/miniconda3/envs/prod/lib/python3.8/site-packages/matplotlib/colorbar.py", line 1249, in _proportional_y
    yscaled = np.ma.filled(norm(yscaled), np.nan)
  File "/opt/miniconda3/envs/prod/lib/python3.8/site-packages/matplotlib/colors.py", line 1245, in __call__
    raise ValueError("minvalue must be less than or equal to maxvalue")
ValueError: minvalue must be less than or equal to maxvalue
```

### Expected outcome

No traceback and colorbar added to the plot.

### Additional information

This code has worked for numerous previous matplotlib releases.

### Operating system

Centos 8 Stream

### Matplotlib Version

3.5.0

### Matplotlib Backend

QtAgg or agg

### Python version

3.8.12

### Jupyter version

n/a

### Installation

conda

### PR DESCRIPTION
**Pull request by jklymak**
## PR Summary

Closes #21671 

Fixes an issue where the BoundaryNorm colorbar was broken if the boundaries were negative.  Added new test to make sure this doesn't break again.  

I'll be honest, I find this part of the colorbar code a bit inscrutable, and it doesn't have a lot of unit tests.  So more bug reports to help get all the combinations down would be helpful.  

## PR Checklist

**Tests and Styling**
- [x] Has pytest style unit tests (and `pytest` passes).
- [x] Is [Flake 8](https://flake8.pycqa.org/en/latest/) compliant (install `flake8-docstrings` and run `flake8 --docstring-convention=all`).

**Documentation**
- [ ] New features are documented, with examples if plot related.
- [ ] New features have an entry in `doc/users/next_whats_new/` (follow instructions in README.rst there).
- [ ] API changes documented in `doc/api/next_api_changes/` (follow instructions in README.rst there).
- [ ] Documentation is sphinx and numpydoc compliant (the docs should [build](https://matplotlib.org/devel/documenting_mpl.html#building-the-docs) without error).

### CODE REVIEW

File: lib/matplotlib/colorbar.py [hunk 1]
```diff
@@ -1150,7 +1150,7 @@ def _mesh(self):
 
     def _forward_boundaries(self, x):
         b = self._boundaries
-        y = np.interp(x, b, np.linspace(0, b[-1], len(b)))
+        y = np.interp(x, b, np.linspace(0, np.abs(b[-1]), len(b)))
```
Comment (anntzer) [inline lib/matplotlib/colorbar.py:1153]: Can you clarify what is being computed here?  Without context, it seems wrong that only the last point of `b` is taken into account, in particular the example in #21671 fails again (with this PR) if `b[-1] == 0` (e.g., `clevs = list(range(-94, 1))`), or rather the resulting plot is wrong.

**Review by anntzer — CHANGES_REQUESTED** [review]
Concerned this is not the right fix, as a variant (above) still fails.

File: lib/matplotlib/colorbar.py (hunk 1 above)
Comment (jklymak) [reply lib/matplotlib/colorbar.py:1153]: Good catch.  Fixed.  OTOH do we know if `clevs=np.arange(1, -9, -1)` should be allowed?

Comment (anntzer) [reply lib/matplotlib/colorbar.py:1153]: I would like decreasing levels to be allowed, but to be clear that's a separate issue that can be dealt with later; the problem here is with increasing levels where the last level is zero.

Comment (jklymak) [reply lib/matplotlib/colorbar.py]: Yes, I can confirm that `clev=np.arange(1, -9, -1)` used to return nonsense on v3.4.3.  It errors on this PR as of my latest push, which seems better than nonsense, but it can probably be made to work.

**Comment by jklymak** [comment]
Just to describe what this does so I'm not just hacking...

`_forward_boundaries` maps from boundaries = [-9, -8, -6, -1, 0] to [0, 2, 4, 6, 8] so that the levels are equally spaced and yielding a colorbar for the `spacing='uniform'` option of `colorbar`; 

cbar

I don't think this needs an image test because we do visually test the boundary norm, just we weren't doing it with negative levels, so the proposed test should be adequate.

File: lib/matplotlib/colorbar.py (hunk 1 above)
Comment (anntzer) [reply lib/matplotlib/colorbar.py]: Would it work to just set the second arg to `range(len(b))` instead?

Comment (jklymak) [reply lib/matplotlib/colorbar.py]: You mean the third arg?

`y = np.interp(x, b, np.arange(len(b)))`

That works as well.

Comment (jklymak) [reply lib/matplotlib/colorbar.py]: The only issue is that there is this 1e-6 scaling in there, that doesn't seem tested, but was apparently there for a reason, so using the actual spread in the boundaries gives a natural scaling for that...

Comment (jklymak) [reply lib/matplotlib/colorbar.py]: though now that I look at it ` y[x > b[-1]+eps] = 2` can't possibly be right...

Comment (anntzer) [reply lib/matplotlib/colorbar.py]: Yes, I meant the third.  I'm also a bit confused by the setting to -1/+2 for out-of-bounds values...

Comment (jklymak) [reply lib/matplotlib/colorbar.py]: No thats clearly wrong.  Fixed in the latest.  However, I played with this, and there are no tests of this happening, so I'm not sure where this hedging came from.  I _think_ it may have been extra ticks in the extend region, but that can't really happen (because its just bounds mapping to bounds) so maybe we don't need all this...

Comment (anntzer) [reply lib/matplotlib/colorbar.py]: I see that in the new version you're mapping to `np.linspace(0, db, len(b))`, which, while not "wrong", still seems weird to me (I still think the mapped-to space should be either `range(len(b))` or `np.linspace(0, 1, len(b))` -- `db` doesn't have any particular meaning in that space).

Comment (jklymak) [reply lib/matplotlib/colorbar.py]: OK, I think that works...

**Comment by jklymak** [comment]
@anntzer This version (boundaries mapped from 0-1) fails on pixel snapping of ticks at `test_contour/test_contourf_log_extension`.  If I change back to map the boundaries between 0-(b[-1]-b[0]) the pixel snapping is the same as on main.  In my opinion the pixel snapping looks better that way, however, I don't think its correct to say its "better".  

Image instability because of pixel snapping is pretty annoying...

**Comment by anntzer** [comment]
Which one is "better" for you? (sorry, that was a bit unclear for me in your message)  I looked at the images; it seems that the difference basically arises because the boundary between two color blocks occurs between two pixel rows, and it's just a matter of deciding whether the tick is drawn on the last row of the first block or the first row of the second block, which is a bit arbitrary?

I think I'd just stick a tol=... on that test (as we *know* where the discrepancy comes from) and be done with it?

**Comment by jklymak** [comment]
Yep - if we just tested PDFs that might be better - bitmapping is done visually on the PDF, rather than with our numerically unstable snapping behaviour in the AGG backends.  However, testing PDFs is slower...

**Comment by anntzer** [comment]
Let's just live with what we have for now? :)

**Comment by jklymak** [comment]
(Yes, but in general I've been favouring png-only tests a lot of time, whereas now I think maybe we should favour pdf-only...  Though pdf sometimes has its own problems.  )

File: lib/matplotlib/tests/test_colorbar.py [hunk 2]
```diff
@@ -873,3 +873,30 @@ def test_proportional_colorbars():
             CS3 = axs[i, j].contourf(X, Y, Z, levels, cmap=cmap, norm=norm,
                                      extend=extends[i])
             fig.colorbar(CS3, spacing=spacings[j], ax=axs[i, j])
+
+
+def test_negative_boundarynorm():
```
Comment (anntzer) [inline lib/matplotlib/tests/test_colorbar.py:878]: spacing between blocks of examples is inconsistent.

**Review by anntzer — APPROVED** [review]
lgtm, modulo CI and possibly squashing.

**Comment by jklymak** [comment]
squashed....

**Review by timhoffm — APPROVED** [review]

**Comment by QuLogic** [comment]
> Yep - if we just tested PDFs that might be better - bitmapping is done visually on the PDF, rather than with our numerically unstable snapping behaviour in the AGG backends. However, testing PDFs is slower...

'visually on the PDF' is just rasterizing at a different step, and is dependent on the rasterizer in use. I'm not sure this is any safer given PDFs sometimes break when updating Ghostscript, even if visually you see no difference.
